modules/loader/src/main/scala/com/snowplowanalytics/snowplow/postgres/loader/Main.scala

```diff
@@ -33,25 +43,64 @@ object Main extends IOApp {
-            for {
-              _ <- loaderConfig.purpose match {
-                case Purpose.Enriched       => utils.prepare[IO](loaderConfig.output.schema, xa)
-                case Purpose.SelfDescribing => IO.unit
-              }
-              badSink = sink.badSink[IO]
-              goodSink = sink.goodSink[IO](UnorderedPipe.forTransactor(xa), state, iglu, processor).andThen(_.through(badSink))
-              s = dataStream.observeEither(badSink, goodSink)
-
-              _ <- s.compile.drain
-            } yield ExitCode.Success
+      case Right(cli) => runWithCli(cli)
+      case Left(error) => IO.delay(logger.error(s"Configuration initialization failure\n$error")).as(ExitCode.Error)
+    }
+
+  def runWithCli(cli: Cli[IO]): IO[ExitCode] = {
+    val res = for {
+      r <- resources.initialize[IO](cli.config.output, cli.iglu)
+      (blocker, xa, state) = r
+      env <- cli.config.input match {
+        case config: Source.Kinesis => KinesisEnv.create[IO](blocker, config, cli.config.monitoring.metrics)
+        case config: Source.PubSub => PubsubEnv.create[IO](blocker, config)
+        case config: Source.LocalFS => LocalEnv.create[IO](blocker, config)
+      }
+    } yield (env, xa, state)
+    res.use { case (env, xa, state) =>
+      runWithEnv(env, cli, xa, state)
+    }
+  }
+
+  private def runWithEnv[A](
+    env: Environment[IO, A],
+    cli: Cli[IO],
+    xa: HikariTransactor[IO],
+    state: State[IO]
+  ): IO[ExitCode] = {
+      val dataStream = env.source
+        .map(record => (record, env.getPayload(record)))
+        .map { case (record, payload) =>
+          val res = payload.leftMap(createError(cli.config.purpose))
+            .flatMap { p =>
+              ParserUtils.parseItem(cli.config.purpose, p)
+            }
+          (record, res)
+        }
+      implicit val db: DB[IO] = DB.interpreter[IO](cli.iglu.resolver, xa, cli.config.output.schema)
+      for {
+        _ <- cli.config.purpose match {
+          case Purpose.Enriched       => utils.prepare[IO](cli.config.output.schema, xa)
+          case Purpose.SelfDescribing => IO.unit
         }
+        s = dataStream.through(sinkAll(OrderedPipe.forTransactor(xa), sink.sinkResult(state, cli.iglu, processor)))
+          .through(env.checkpointer)
+        _ <- s.compile.drain
+      } yield ExitCode.Success
+  }
 
-      case Left(error) =>
-        IO.delay(logger.error(s"Configuration initialization failure\n$error")).as(ExitCode.Error)
+  private def sinkAll[A](orderedPipe: OrderedPipe[IO], sinkResult: Either[BadData, Data] => IO[Unit]): Pipe[IO, (A, Either[BadData, Data]), A] =
+    orderedPipe {
+      case (record, payload) => sinkResult(payload).map(_ => record)
     }
+
+  private def createError(kind: Purpose)(payload: String): BadData =
+    kind match {
+      case Purpose.Enriched =>
+        val badRow = BadRow.LoaderParsingError(Cli.processor, NotTSV, Payload.RawPayload(payload))
```

Comment:
Just in case for `NotTSV`, enriched event can be TSV but still not being able to be parsed, e.g. in case of incorrect field format.

Reply:
Actually, this function is creating error for kinesis only and when you look at there, error is only possible if there is problem in string decoding which means that item is not TSV. In order to reflect this situtation better, I moved this functionality to `KinesisEnv`. Let me know if you see any problem with the new approach.